AtomLinter/linter-eslint
# Ignore home config from disableWhenNoEslintConfig

Fixes #773

## Description

This PR prevents an eslint config file located in a user's home directory (e.g. `/users/person/.eslintrc`) from enabling linting of projects within the home directory when the setting `Disable when no ESLint config is found` is enabled.  

Note, because of the way we traverse up the directory structure looking for config files, projects outside of a user's home do not currently find a `fallback` config in a user's home, and thus those projects are already being disabled correctly.

## Notes

This adds a dependency on `user-home`, which is used to detect the user's home directory and return it as a string.  This result is compared to the directory containing the config file returned from the config file search, to determine if the file should be treated as a valid project config file or a fallback.

Also, I've added a few specs here around the `disableWhenNoEslintConfig` setting, but I wasn't able to create a good red-green test for the particular fix in this PR.  That's because I could not find a way to mock the functionality of `findCached` in such a way to fake the finding of a config in the home directory.  My plan was to put a file with linting errors in the OS temp dir, (not under the home dir), mock out `findCached` to make it seem there was a config file in the user home, and then show that with this PR, linting is still disabled.  However, my go-to tool for such purposes, `proxyquire` does not seem to work correctly with electron's `require` implementation.  If you know if any other way to mock dependencies in the way I'm attempting, I'd be keen to hear it.

## How to test
Despite not having a good automated test for this change, it's fairly easy to check manually.

- Put a `.js` file similar to the `badInline.js` file added in this pr, somewhere in your home directory, like `~/experiment/foo.js`.
- Assuming you don't already have an eslint config in your home, `touch ~/.eslintrc`
- Verify that without the changes in this PR, the setting for `disableWhenNoEslintConfig` has no effect, and linting is performed on the file no matter the setting.
- Then, with the changes in this PR, verify that linting is indeed disabled when (and only when) `disableWhenNoEslintConfig` is turned on.

## Warning

The base branch of this PR is not master, but rather another branch with an open PR.  I have a good feeling that other PR (#777) will be merged, and rather than rebase this branch afterwards, I've just based on that, and will move the base to master after #777 is merged.


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/atomlinter/linter-eslint/778)
